Documentation/teaching/lectures/smp.rst

```diff
@@ -0,0 +1,1156 @@
+the load and store operations are interleaved across CPUs, like in the
+example below where incrementing a value from two processors will
+produce an unexpected result:
+
+.. slide:: Atomic operations may not be atomic on SMP systems
+   :inline-contents: True
+   :level: 2
+
+   |_|
+
+   .. ditaa::
+
+
+                                   +------------+
+                                   |   Memory   |
+     +-------------+   LOAD (0)    |            |               +-------------+
+     |    CPU 0    |<--------------|   v <- 0   |    LOAD (0)   |    CPU 1    |
+     |             |   STORE (1)   |            |-------------->|             |
+     |    inc v    |-------------->|   v <- 1   |    STORE (1)  |    inc v    |
+     | cEEE        |               |   v <- 1   |<--------------| cEEE        |
+     +-------------+               | cEEE       |               +-------------+
+                                   +------------+
+
+
+In order to provide atomic operations on SMP systems different
+architectures use different techniques. For example, on x86 a LOCK
+prefix is used to lock the system bus while executing the prefixed
+operation:
+
+.. slide:: Fixing atomic operations for SMP systems (x86)
+   :inline-contents: True
+   :level: 2
+
+   |_|
+
+   .. ditaa::
+
+                                   +------------+
+     +-------------+   BUS LOCK    |   Memory   |
+     |    CPU 1    |<------------->|            |
+     |             |   LOAD (0)    |            |
+     |    inc v    |<--------------|   v <- 0   |
+     |             |   STORE (1)   |            |
+     |             |-------------->|   v <- 1   |
+     |             |  BUS UNLOCK   |            |
+     | cEEE        |<------------->|            |   BUS LOCK    +-------------+
+     +-------------+               |            |<------------->|    CPU 1    |
+                                   |            |   LOAD (1)    |             |
+                                   |            |<--------------|    inc v    |
+                                   |   v <- 2   |   STORE (2)   |             |
+                                   |            |-------------->|             |
+                                   |            |  BUS UNLOCK   |             |
+                                   | cEEE       |<------------->| cEEE        |
+                                   +------------+               +-------------+
+
+
+On ARM the LDREX and STREX instructions are used together to guarantee
+atomic access: LDREX loads a value and signals the exclusive monitor
+that an atomic operation is in progress. The STREX attempts to store a
+new value but only succeeds if the exclusive monitor has not detected
+other exclusive operations. So, to implement atomic operations the
+programmer must retry the operation (both LDREX and STREX) until the
+exclusive monitor signals a success.
+
+As seen from the implementation details, atomic operations are
```

Comment:
This is the case with all synchronization/exclusive access mechanisms. Do we need to stress this for atomic operations?

Reply:
Yes, atomic operations are often seen as light or efficient alternatives to locking. I added more context:

```
Although they are often interpreted as "light" or "efficient"
synchronization mechanisms (because they "don't require spinning or
context switches", or because they "are implemented in hardware so
they must be more efficient", or because they "are just instructions
so they must have similar efficiency as other instructions"), as seen
from the implementation details, atomic operations are expensive and
they must be avoided when possible.
```